src/iosrv/drpc_ras.c

```diff
@@ -0,0 +1,316 @@
+	 daos_obj_id_t *oid, char *cop, Shared__RASEvent *evt)
+{
+	struct timeval	 tv;
+	struct tm	*tm;
+	int		 rc;
+
+	evt->timestamp = NULL;
+	evt->hostname = NULL;
+	evt->msg = NULL;
+	/* max value indicates nil rank */
+	evt->rank = UINT32_MAX;
+	evt->hw_id = NULL;
+	evt->proc_id = NULL;
+	evt->thread_id = NULL;
+	evt->job_id = NULL;
+	evt->pool_uuid = NULL;
+	evt->cont_uuid = NULL;
+	evt->obj_id = NULL;
+	evt->ctl_op = NULL;
+
+	/* Populate mandatory RAS fields. */
+
+	(void)gettimeofday(&tv, 0);
+	tm = localtime(&tv.tv_sec);
+	if (tm)
+		D_ASPRINTF(evt->timestamp,
+			   "%04d/%02d/%02d-%02d:%02d:%02d.%02ld",
+			   tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
+			   tm->tm_hour, tm->tm_min, tm->tm_sec,
+			   (long)tv.tv_usec / 10000);
+
+	evt->id = (uint32_t)id;
+	evt->type = (uint32_t)type;
+	evt->severity = (uint32_t)sev;
+
+	if (strnlen(dss_hostname, DSS_HOSTNAME_MAX_LEN) == 0) {
+		D_ERROR("missing hostname parameter\n");
+		D_GOTO(out_fail, rc = -DER_UNINIT);
+	}
+	D_STRNDUP(evt->hostname, dss_hostname, DSS_HOSTNAME_MAX_LEN);
+
+	if (!msg || strnlen(msg, DAOS_RAS_STR_FIELD_SIZE) == 0) {
+		D_ERROR("missing msg parameter\n");
+		D_GOTO(out_fail, rc = -DER_INVAL);
+	}
+	evt->msg = msg;
+
+	/* Populate optional RAS fields. */
+
+	if (hid)
+		evt->hw_id = hid;
+	if (rank)
+		evt->rank = (uint32_t)*rank;
+	if (jid)
+		evt->job_id = jid;
+	if (puuid && !uuid_is_null(*puuid))
+		D_ASPRINTF(evt->pool_uuid, DF_UUIDF, DP_UUID(*puuid));
+	if (cuuid && !uuid_is_null(*cuuid))
+		D_ASPRINTF(evt->cont_uuid, DF_UUIDF, DP_UUID(*cuuid));
+	if (oid)
+		D_ASPRINTF(evt->obj_id, DF_OID, DP_OID(*oid));
+	if (cop)
+		evt->ctl_op = cop;
+
+	return 0;
+out_fail:
+	free_ras(evt);
+
+	return rc;
+}
+
+static void
+log_ras(Shared__RASEvent *evt)
+{
+	FILE		*stream;
+	char		*buf = NULL;
+	size_t		 len;
+
+	stream = open_memstream(&buf, &len);
+	if (!stream) {
+		D_ERROR("open_memstream failed: "DF_RC"\n", DP_RC(-DER_NOMEM));
+		return;
+	}
+
+	/* Log mandatory RAS fields. */
+
+	D_FPRINTF(stream, " id: [%s]", ras_event2str(evt->id));
+	if (evt->timestamp)
+		D_FPRINTF(stream, " ts: [%s]", evt->timestamp);
+	if (evt->hostname)
+		D_FPRINTF(stream, " host: [%s]", evt->hostname);
+	D_FPRINTF(stream, " type: [%s] sev: [%s]", ras_type2str(evt->type),
+		  ras_sev2str(evt->severity));
+	if (evt->msg)
+		D_FPRINTF(stream, " msg: [%s]", evt->msg);
+
+	/* Log optional RAS fields. */
+
+	if (evt->hw_id)
+		D_FPRINTF(stream, " hwid: [%s]", evt->hw_id);
+	if (evt->rank != UINT32_MAX)
+		D_FPRINTF(stream, " rank: [%u]", evt->rank);
+	if (evt->job_id)
+		D_FPRINTF(stream, " jobid: [%s]", evt->job_id);
+	if (evt->pool_uuid)
+		D_FPRINTF(stream, " puuid: [%s]", evt->pool_uuid);
+	if (evt->cont_uuid)
+		D_FPRINTF(stream, " cuuid: [%s]", evt->cont_uuid);
+	if (evt->obj_id)
+		D_FPRINTF(stream, " oid: [%s]", evt->obj_id);
+	if (evt->ctl_op)
+		D_FPRINTF(stream, " control_op: [%s]", evt->ctl_op);
+	if (evt->extended_info_case == \
```

Comment:
(style)  Avoid unnecessary line continuations

Reply:
Should be removed.